src/main/java/com/powsybl/balances_adjustment/util/ControlArea.java

```diff
@@ -0,0 +1,109 @@
+        terminalsAndBoundaries.addAll(cgmesControlAreaMapping.getCgmesControlArea(controlAreaId).getBoundaries());
+
+        voltageLevelIdsCache = getVoltageLevelIds(controlAreaId, network, terminalsAndBoundaries);
+    }
+
+    @Override
+    public double getNetPosition() {
+        return terminalsAndBoundaries.parallelStream().mapToDouble(o -> {
+            if (o instanceof Terminal) {
+                return getLeavingFlow((Terminal) o);
+            } else if (o instanceof Boundary) {
+                return -((Boundary) o).getP();
+            }
+            throw new AssertionError();
+        }).sum();
+    }
+
+    @Override
+    public Collection<String> getVoltageLevelIds() {
+        return Collections.unmodifiableSet(voltageLevelIdsCache);
+    }
+
+    private static Set<String> getVoltageLevelIds(String controlAreaId, Network network, Set<Object> terminalsAndBoundaries) {
+        Graph<Identifiable, Object> voltageLevelGraph = createVoltageLevelGraph(network, terminalsAndBoundaries);
+        List<Set<Identifiable>> connectedSets = new ConnectivityInspector<>(voltageLevelGraph)
+                .connectedSets()
+                .stream()
+                .filter(set -> terminalsAndBoundaries.stream().anyMatch(o -> set.contains(getVoltageLevel(o))))
+                .collect(Collectors.toList());
+        if (connectedSets.isEmpty()) {
+            throw new PowsyblException("Control area " + controlAreaId + " does not contain any voltage level");
+        }
+        if (connectedSets.size() > 1) {
+            throw new PowsyblException("Control area " + controlAreaId + " contains several connected sets of voltage levels");
+        }
+        return connectedSets.iterator().next()
+                .stream()
+                .filter(o -> o instanceof VoltageLevel)
+                .map(Identifiable::getId)
+                .collect(Collectors.toSet());
+    }
+
+    private static Graph<Identifiable, Object> createVoltageLevelGraph(Network network, Set<Object> terminalsAndBoundaries) {
+        Graph<Identifiable, Object> voltageLevelGraph = new Pseudograph<>(Identifiable.class);
+        network.getVoltageLevelStream().forEach(voltageLevelGraph::addVertex);
+        network.getBranchStream()
+                .filter(b -> !terminalsAndBoundaries.contains(b.getTerminal1()) && !terminalsAndBoundaries.contains(b.getTerminal2()))
+                .filter(b -> {
+                    if (b instanceof TieLine) {
+                        TieLine tl = (TieLine) b;
+                        return !terminalsAndBoundaries.contains(tl.getHalf1().getBoundary()) && !terminalsAndBoundaries.contains(tl.getHalf2().getBoundary());
+                    }
+                    return true;
+                })
+                .forEach(b -> voltageLevelGraph.addEdge(b.getTerminal1().getVoltageLevel(), b.getTerminal2().getVoltageLevel(), b));
```

Comment:
you can remove the last argument of `addEdge()`, as we don't need the branch object on the edge.